File: src/treemapper/diffctx/edges/semantic/dart.py
from __future__ import annotations

import re
from collections import defaultdict
from pathlib import Path

from ...config.weights import EDGE_WEIGHTS
from ...types import Fragment, FragmentId
from ..base import EdgeBuilder, EdgeDict, FragmentIndex, discover_files_by_refs

_DART_EXTS = {".dart"}

_IMPORT_RE = re.compile(r"^\s*import\s+['\"]([^'\"]{1,300})['\"]", re.MULTILINE)
_EXPORT_RE = re.compile(r"^\s*export\s+['\"]([^'\"]{1,300})['\"]", re.MULTILINE)
_PART_RE = re.compile(r"^\s*part\s+['\"]([^'\"]{1,300})['\"]", re.MULTILINE)
_PART_OF_RE = re.compile(r"^\s*part\s+of\s+['\"]([^'\"]{1,300})['\"]", re.MULTILINE)

_CLASS_RE = re.compile(r"^\s*(?:abstract\s+)?class\s+(\w+)", re.MULTILINE)
_MIXIN_RE = re.compile(r"^\s*mixin\s+(\w+)", re.MULTILINE)
_ENUM_RE = re.compile(r"^\s*enum\s+(\w+)", re.MULTILINE)
_TYPEDEF_RE = re.compile(r"^\s*typedef\s+(\w+)", re.MULTILINE)
_EXTENSION_RE = re.compile(r"^\s*extension\s+(\w+)\s+on\s+(\w+)", re.MULTILINE)
_TOP_FUNC_RE = re.compile(r"^(?:(?:[\w<>?,]+\s+)+)?([a-z_]\w*)\s*\(", re.MULTILINE)

_EXTENDS_RE = re.compile(r"\bextends\s+(\w+)")
_IMPLEMENTS_RE = re.compile(r"\bimplements\s+([^{]+)")
_WITH_RE = re.compile(r"\bwith\s+([^{]+)")

_TYPE_REF_RE = re.compile(r"(?<![a-z_])([A-Z]\w{1,100})\b")
_FUNC_CALL_RE = re.compile(r"(?<!\w)([a-z_]\w{1,100})\s*\(")

_DART_KEYWORDS = frozenset(
    {
        "if",
        "else",
        "for",
        "while",
        "do",
        "switch",
        "case",
        "return",
        "break",
        "continue",
        "throw",
        "try",
        "catch",
        "finally",
        "assert",
        "new",
        "const",
        "var",
        "final",
        "void",
        "dynamic",
        "class",
        "abstract",
        "extends",
        "implements",
        "with",
        "mixin",
        "enum",
        "typedef",
        "import",
        "export",
        "part",
        "library",
        "show",
        "hide",
        "as",
        "is",
        "in",
        "this",
        "super",
        "null",
        "true",
        "false",
        "async",
        "await",
        "yield",
        "print",
        "main",
        "get",
        "set",
    }
)

_DART_COMMON_TYPES = frozenset(
    {
        "String",
        "int",
        "double",
        "bool",
        "num",
        "List",
        "Map",
        "Set",
        "Future",
        "Stream",
        "Iterable",
        "Object",
        "Function",
        "Type",
        "Duration",
        "DateTime",
        "Null",
        "Never",
        "void",
        "dynamic",
    }
)

_DIFF_IMPORT_RE = re.compile(r"^\+\s*import\s+['\"]([^'\"]{1,300})['\"]", re.MULTILINE)
_DIFF_EXPORT_RE = re.compile(r"^\+\s*export\s+['\"]([^'\"]{1,300})['\"]", re.MULTILINE)
_DIFF_PART_RE = re.compile(r"^\+\s*part\s+['\"]([^'\"]{1,300})['\"]", re.MULTILINE)


def _is_dart_file(path: Path) -> bool:
    return path.suffix.lower() in _DART_EXTS


def _extract_refs(content: str) -> tuple[set[str], set[str], set[str]]:
    imports: set[str] = set()
    exports: set[str] = set()
    parts: set[str] = set()

    for m in _IMPORT_RE.finditer(content):
        imports.add(m.group(1))
    for m in _EXPORT_RE.finditer(content):
        exports.add(m.group(1))
    for m in _PART_RE.finditer(content):
        parts.add(m.group(1))
    for m in _PART_OF_RE.finditer(content):
        parts.add(m.group(1))

    return imports, exports, parts


def _extract_definitions(content: str) -> tuple[set[str], set[str]]:
    types: set[str] = set()
    types.update(m.group(1) for m in _CLASS_RE.finditer(content))
    types.update(m.group(1) for m in _MIXIN_RE.finditer(content))
    types.update(m.group(1) for m in _ENUM_RE.finditer(content))
    types.update(m.group(1) for m in _TYPEDEF_RE.finditer(content))
    for m in _EXTENSION_RE.finditer(content):
        types.add(m.group(1))

    funcs: set[str] = set()
    for m in _TOP_FUNC_RE.finditer(content):
        name = m.group(1)
        if name not in _DART_KEYWORDS and len(name) >= 2:
            funcs.add(name)

    return funcs, types


def _extract_inheritance(content: str) -> set[str]:
    parents: set[str] = set()
    for m in _EXTENDS_RE.finditer(content):
        parents.add(m.group(1))
    for m in _IMPLEMENTS_RE.finditer(content):
        for part in m.group(1).split(","):
            name = part.strip().split("<")[0].strip()
            if name and name[0].isupper():
                parents.add(name)
    for m in _WITH_RE.finditer(content):
        for part in m.group(1).split(","):
            name = part.strip().split("<")[0].strip()
            if name and name[0].isupper():
                parents.add(name)
    return parents


def _extract_references(content: str) -> tuple[set[str], set[str]]:
    type_refs = {m.group(1) for m in _TYPE_REF_RE.finditer(content) if m.group(1) not in _DART_COMMON_TYPES}
    func_calls = {m.group(1) for m in _FUNC_CALL_RE.finditer(content) if m.group(1) not in _DART_KEYWORDS}
    return type_refs, func_calls


def _ref_to_filename(ref: str) -> str:
    name = ref.split("/")[-1].lower()
    if name.startswith("package:"):
        name = name.split(":")[-1]
    return name


class DartEdgeBuilder(EdgeBuilder):
    weight = 0.65
    import_weight = EDGE_WEIGHTS["dart_import"].forward
    export_weight = EDGE_WEIGHTS["dart_export"].forward
    type_weight = EDGE_WEIGHTS["dart_type"].forward
    fn_weight = EDGE_WEIGHTS["dart_fn"].forward
    inheritance_weight = EDGE_WEIGHTS["dart_inheritance"].forward
    reverse_weight_factor = EDGE_WEIGHTS["dart_import"].reverse_factor

    def needs_from_diff(self, diff_content: str) -> list[str]:
        refs: list[str] = []
        for m in _DIFF_IMPORT_RE.finditer(diff_content):
            refs.append(_ref_to_filename(m.group(1)))
        for m in _DIFF_EXPORT_RE.finditer(diff_content):
            refs.append(_ref_to_filename(m.group(1)))
        for m in _DIFF_PART_RE.finditer(diff_content):
            refs.append(_ref_to_filename(m.group(1)))
        return refs

    def discover_related_files(
        self,
        changed_files: list[Path],
        all_candidate_files: list[Path],
        repo_root: Path | None = None,
        **kwargs: object,
    ) -> list[Path]:
        dart_changed = [f for f in changed_files if _is_dart_file(f)]
        if not dart_changed:
            return []

        refs: set[str] = set()
        for f in dart_changed:
            try:
                content = f.read_text(encoding="utf-8")
                imports, exports, parts = _extract_refs(content)
                for r in imports | exports | parts:
                    refs.add(_ref_to_filename(r))
            except (OSError, UnicodeDecodeError):
                continue

        return discover_files_by_refs(refs, changed_files, all_candidate_files, repo_root)

    def build(self, fragments: list[Fragment], repo_root: Path | None = None) -> EdgeDict:
        dart_frags = [f for f in fragments if _is_dart_file(f.path)]
        if not dart_frags:
            return {}

        edges: EdgeDict = {}
        idx = FragmentIndex(fragments, repo_root)
        type_defs, fn_defs = self._build_indices(dart_frags)

        for df in dart_frags:
            self._add_fragment_edges(df, idx, type_defs, fn_defs, edges)

        return edges

    def _build_indices(self, dart_frags: list[Fragment]) -> tuple[dict[str, list[FragmentId]], dict[str, list[FragmentId]]]:
        type_defs: dict[str, list[FragmentId]] = defaultdict(list)
        fn_defs: dict[str, list[FragmentId]] = defaultdict(list)

        for f in dart_frags:
            funcs, types = _extract_definitions(f.content)
            for t in types:
                type_defs[t.lower()].append(f.id)
            for fn in funcs:
                fn_defs[fn.lower()].append(f.id)

        return type_defs, fn_defs

    def _link_imports_exports_parts(
        self,
        df: Fragment,
        idx: FragmentIndex,
        edges: EdgeDict,
    ) -> None:
        imports, exports, parts = _extract_refs(df.content)
        for ref in imports:
            self._link_ref(df.id, ref, idx, edges, self.import_weight)
        for ref in exports:
            self._link_ref(df.id, ref, idx, edges, self.export_weight)
        for ref in parts:
            self._link_ref(df.id, ref, idx, edges, self.import_weight)

    def _link_inheritance(
        self,
        df: Fragment,
        type_defs: dict[str, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        for parent in _extract_inheritance(df.content):
            for fid in type_defs.get(parent.lower(), []):
                if fid != df.id:
                    self.add_edge(edges, df.id, fid, self.inheritance_weight)

    def _link_type_refs(
        self,
        df: Fragment,
        type_defs: dict[str, list[FragmentId]],
        self_type_lower: set[str],
        edges: EdgeDict,
    ) -> None:
        type_refs, _ = _extract_references(df.content)
        for type_ref in type_refs:
            if type_ref.lower() in self_type_lower:
                continue
            for fid in type_defs.get(type_ref.lower(), []):
                if fid != df.id:
                    self.add_edge(edges, df.id, fid, self.type_weight)

    def _link_fn_calls(
        self,
        df: Fragment,
        fn_defs: dict[str, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        _, func_calls = _extract_references(df.content)
        for func_call in func_calls:
            for fid in fn_defs.get(func_call.lower(), []):
                if fid != df.id:
                    self.add_edge(edges, df.id, fid, self.fn_weight)

    def _add_fragment_edges(
        self,
        df: Fragment,
        idx: FragmentIndex,
        type_defs: dict[str, list[FragmentId]],
        fn_defs: dict[str, list[FragmentId]],
        edges: EdgeDict,
    ) -> None:
        self._link_imports_exports_parts(df, idx, edges)
        self._link_inheritance(df, type_defs, edges)

        _, self_types = _extract_definitions(df.content)
        self_type_lower = {t.lower() for t in self_types}

        self._link_type_refs(df, type_defs, self_type_lower, edges)
        self._link_fn_calls(df, fn_defs, edges)

    def _link_ref(
        self,
        src_id: FragmentId,
        ref: str,
        idx: FragmentIndex,
        edges: EdgeDict,
        weight: float,
    ) -> None:
        filename = _ref_to_filename(ref)
        for name, frag_ids in idx.by_name.items():
            if name == filename or name == filename.replace(".dart", ""):
                for fid in frag_ids:
                    if fid != src_id:
                        self.add_edge(edges, src_id, fid, weight)
                        return

        self.link_by_path_match(src_id, ref, idx, edges, weight)

File: diffctx/src/edges/semantic/dart.rs
use std::path::{Path, PathBuf};

use once_cell::sync::Lazy;
use regex::Regex;
use rustc_hash::{FxHashMap, FxHashSet};

use crate::config::weights::EDGE_WEIGHTS;
use crate::types::{Fragment, FragmentId};

use super::super::base::{self, EdgeBuilder, FragmentIndex, add_edge, discover_files_by_refs, link_by_name};
use super::super::EdgeDict;

static DART_EXTENSIONS: Lazy<FxHashSet<&str>> =
    Lazy::new(|| [".dart"].iter().copied().collect());

fn is_dart_file(path: &Path) -> bool {
    let ext = base::file_ext(path);
    DART_EXTENSIONS.contains(ext.as_str())
}

static IMPORT_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r#"(?m)^\s*import\s+['"]([^'"]+)['"]"#).unwrap());
static EXPORT_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r#"(?m)^\s*export\s+['"]([^'"]+)['"]"#).unwrap());
static PART_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r#"(?m)^\s*part\s+['"]([^'"]+)['"]"#).unwrap());
static PART_OF_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r#"(?m)^\s*part\s+of\s+['"]([^'"]+)['"]"#).unwrap());
static CLASS_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?m)^\s*(?:abstract\s+)?class\s+(\w+)").unwrap());
static MIXIN_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?m)^\s*mixin\s+(\w+)").unwrap());
static EXTENSION_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?m)^\s*extension\s+(\w+)").unwrap());
static FUNC_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?m)^\s*(?:\w+\s+)*(\w+)\s*[<(]").unwrap());
static TYPE_REF_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"\b([A-Z]\w+)\b").unwrap());
static CALL_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"\b([a-z_]\w+)\s*\(").unwrap());

static DART_KEYWORDS: Lazy<FxHashSet<&str>> = Lazy::new(|| {
    [
        "if", "else", "for", "while", "do", "switch", "case", "break", "continue",
        "return", "var", "final", "const", "void", "null", "true", "false", "new",
        "this", "super", "class", "extends", "implements", "with", "abstract",
        "import", "export", "library", "part", "typedef", "enum", "mixin",
        "extension", "async", "await", "yield", "try", "catch", "finally",
        "throw", "rethrow", "assert", "in", "is", "as", "dynamic", "Function",
        "String", "int", "double", "bool", "List", "Map", "Set", "Future",
        "Stream", "Iterable", "Object", "Null", "Never", "Type", "print",
    ]
    .iter()
    .copied()
    .collect()
});

fn extract_refs(content: &str) -> FxHashSet<String> {
    let mut refs = FxHashSet::default();
    for cap in IMPORT_RE.captures_iter(content) {
        refs.insert(cap[1].to_string());
    }
    for cap in EXPORT_RE.captures_iter(content) {
        refs.insert(cap[1].to_string());
    }
    for cap in PART_RE.captures_iter(content) {
        refs.insert(cap[1].to_string());
    }
    for cap in PART_OF_RE.captures_iter(content) {
        refs.insert(cap[1].to_string());
    }
    refs
}

fn extract_defines(content: &str) -> FxHashSet<String> {
    let mut defs = FxHashSet::default();
    for cap in CLASS_RE.captures_iter(content) {
        defs.insert(cap[1].to_string());
    }
    for cap in MIXIN_RE.captures_iter(content) {
        defs.insert(cap[1].to_string());
    }
    for cap in EXTENSION_RE.captures_iter(content) {
        defs.insert(cap[1].to_string());
    }
    for cap in FUNC_RE.captures_iter(content) {
        let name = &cap[1];
        if !DART_KEYWORDS.contains(name) {
            defs.insert(name.to_string());
        }
    }
    defs
}

fn extract_type_refs(content: &str) -> FxHashSet<String> {
    TYPE_REF_RE
        .captures_iter(content)
        .map(|c| c[1].to_string())
        .filter(|n| !DART_KEYWORDS.contains(n.as_str()))
        .collect()
}

fn extract_calls(content: &str) -> FxHashSet<String> {
    CALL_RE
        .captures_iter(content)
        .map(|c| c[1].to_string())
        .filter(|n| !DART_KEYWORDS.contains(n.as_str()))
        .collect()
}

pub struct DartEdgeBuilder;

impl EdgeBuilder for DartEdgeBuilder {
    fn build(&self, fragments: &[Fragment], _repo_root: Option<&Path>) -> EdgeDict {
        let dart_frags: Vec<&Fragment> = fragments
            .iter()
            .filter(|f| is_dart_file(Path::new(f.path())))
            .collect();
        if dart_frags.is_empty() {
            return FxHashMap::default();
        }

        let import_weight = EDGE_WEIGHTS["dart_import"].forward;
        let type_weight = EDGE_WEIGHTS["dart_type"].forward;
        let fn_weight = EDGE_WEIGHTS["dart_fn"].forward;
        let inheritance_weight = EDGE_WEIGHTS["dart_inheritance"].forward;
        let reverse_factor = EDGE_WEIGHTS["dart_import"].reverse_factor;

        let idx = FragmentIndex::new(fragments, _repo_root);

        let mut name_to_defs: FxHashMap<String, Vec<FragmentId>> = FxHashMap::default();
        let mut frag_defines: FxHashMap<FragmentId, FxHashSet<String>> = FxHashMap::default();

        for f in &dart_frags {
            let defs = extract_defines(&f.content);
            for name in &defs {
                name_to_defs.entry(name.clone()).or_default().push(f.id.clone());
            }
            frag_defines.insert(f.id.clone(), defs);
        }

        let mut edges: EdgeDict = FxHashMap::default();

        for f in &dart_frags {
            let self_defs = frag_defines.get(&f.id).cloned().unwrap_or_default();

            let file_refs = extract_refs(&f.content);
            for r in &file_refs {
                link_by_name(&f.id, r, &idx, &mut edges, import_weight, reverse_factor);
            }

            let type_refs = extract_type_refs(&f.content);
            for name in &type_refs {
                if self_defs.contains(name) {
                    continue;
                }
                if let Some(dst_ids) = name_to_defs.get(name) {
                    for dst_id in dst_ids {
                        if dst_id != &f.id {
                            add_edge(&mut edges, &f.id, dst_id, type_weight, reverse_factor);
                        }
                    }
                }
            }

            let calls = extract_calls(&f.content);
            for name in &calls {
                if self_defs.contains(name) {
                    continue;
                }
                if let Some(dst_ids) = name_to_defs.get(name) {
                    for dst_id in dst_ids {
                        if dst_id != &f.id {
                            add_edge(&mut edges, &f.id, dst_id, fn_weight, reverse_factor);
                        }
                    }
                }
            }

            for ident in &f.identifiers {
                if self_defs.contains(ident) {
                    continue;
                }
                if let Some(dst_ids) = name_to_defs.get(ident) {
                    for dst_id in dst_ids {
                        if dst_id != &f.id {
                            add_edge(&mut edges, &f.id, dst_id, inheritance_weight, reverse_factor);
                        }
                    }
                }
            }
        }

        edges
    }

    fn discover_related_files(
        &self,
        changed: &[PathBuf],
        candidates: &[PathBuf],
        repo_root: Option<&Path>,
        file_cache: Option<&FxHashMap<PathBuf, String>>,
    ) -> Vec<PathBuf> {
        let dart_changed: Vec<&PathBuf> = changed.iter().filter(|f| is_dart_file(f)).collect();
        if dart_changed.is_empty() {
            return vec![];
        }

        let mut all_refs = FxHashSet::default();
        for f in &dart_changed {
            let content = base::read_file_cached(f, file_cache);
            if let Some(c) = content {
                all_refs.extend(extract_refs(&c));
            }
        }

        discover_files_by_refs(&all_refs, changed, candidates, repo_root)
    }
}
